README.md

```diff
@@ -1,367 +1,654 @@
+### Payload
 
-#### How can I disable or customize automatic padding?
-You can customize automatic padding in order to better fit your needs.
+Sending notifications is nice, but you may need to send a payload to customize messages.
+The payload may be a string, or a JSON object. The structure of the later is described in the next section.
 
-Here are some ideas of settings:
-* (default) `Encryption::MAX_COMPATIBILITY_PAYLOAD_LENGTH` (3052 bytes) for compatibility purposes with Firefox for Android
-* `Encryption::MAX_PAYLOAD_LENGTH` (4078 bytes) for maximum security
-* `false` for maximum performance
-* If you know your payloads will not exceed `X` bytes, then set it to `X` for the best balance between security and performance.
+The payload is encrypted on server side and decrypted by the browser.
+To do so, you shall add Content Encoding services.
+The specification defines two encodings that are very similar: `aesgcm` and `aes128gcm`.
+Please use both of them.
 
 ```php
 <?php
 
-use Minishlink\WebPush\WebPush;
+use Minishlink\WebPush\ExtensionManager;
+use Minishlink\WebPush\Notification;
+use Minishlink\WebPush\Payload\PayloadExtension;
+use Minishlink\WebPush\Payload\AES128GCM;
+use Minishlink\WebPush\Payload\AESGCM;
+
+$cache = '…'; // PSR-6 service
+
+$aes128gcm = AES128GCM::create()
+    ->setCache($cache) // Optional, but highly recommended
+;
+
+$aesgcm = AESGCM::create()
+    ->setCache($cache) // Optional, but highly recommended
+;
+
+$payloadExtension = PayloadExtension::create()
+    ->addContentEncoding($aes128gcm)
+    ->addContentEncoding($aesgcm)
+;
 
-$webPush = new WebPush();
-$webPush->setAutomaticPadding(false); // disable automatic padding
-$webPush->setAutomaticPadding(512); // enable automatic padding to 512 bytes (you should make sure that your payload is less than 512 bytes, or else an attacker could guess the content)
-$webPush->setAutomaticPadding(true); // enable automatic padding to default maximum compatibility length
+$extensionManager = ExtensionManager::create()
+    ->add($payloadExtension);
+;
+
+$notification = Notification::create()
+    ->withPayload('Hello World!')
+;
 ```
 
-### Customizing the HTTP client
-WebPush uses [Guzzle](https://github.com/guzzle/guzzle). It will use the most appropriate client it finds, 
-and most of the time it will be `MultiCurl`, which allows to send multiple notifications in parallel.
+#### Padding
+
+By default, both `AESGCM` and `AES128GCM` content encodings use the recommended padding.
+This feature is used to prevent attacks (Size and Timing) that could reveal the notification payload.
+
+You can change the padding size if needed using one of the three options:
+
+* No padding (not recommended)
+* Recommended padding (default)
+* Maximum padding
+* Custom padding
+
+With the maximum padding, the encrypted payload will always have a size of 4096 bytes.
+This may increase the computation time, that’s why this padding is not enabled, but highly recommended.
```

Comment:
The reason of the existence of this recommended padding is because Firefox for mobile doesn't like when there is a too big message

Reply:
OK. I haven’t found any issue for now with the maximum padding.
Do you think we should remove the recommended one?